python/sglang/srt/managers/controller/infer_batch.py

```diff
@@ -947,28 +945,56 @@ def init_flashinfer_args(
     num_qo_heads = model_runner.model_config.num_attention_heads // model_runner.tp_size
-    num_kv_heads = model_runner.model_config.get_num_kv_heads(model_runner.tp_size)
+    # NOTE (yifan): we partitioned K and V along both TP and SP dimensions.
+    # And here tp_size represents actual TP size * SP size.
+    num_kv_heads = model_runner.model_config.get_num_kv_heads(
+        model_runner.tp_size // model_runner.sp_size
+    )
     head_dim = model_runner.model_config.head_dim
     batch_size = len(req_pool_indices)
 
     if forward_mode == ForwardMode.DECODE:
         paged_kernel_lens = seq_lens
     else:
+        # FIXME (yifan): Simplify the layout adjust code below.
+        seq_lens_cpu = seq_lens.cpu().numpy()
+        prefix_lens_cpu = prefix_lens.cpu().numpy()
+        for i in range(batch_size):
+            seq_len = seq_lens_cpu[i]
+            prefix_len = prefix_lens_cpu[i]
+            seq_lens_cpu[i] = seq_len // model_runner.sp_size + (
+                seq_len % model_runner.sp_size > model_runner.sp_rank
+            )
+            prefix_lens_cpu[i] = prefix_len // model_runner.sp_size + (
+                prefix_len % model_runner.sp_size > model_runner.sp_rank
+            )
+        seq_lens = torch.tensor(seq_lens_cpu, dtype=torch.int32, device="cuda")
+        prefix_lens = torch.tensor(prefix_lens_cpu, dtype=torch.int32, device="cuda")
         paged_kernel_lens = prefix_lens
 
     kv_indptr = torch.zeros((batch_size + 1,), dtype=torch.int32, device="cuda")
     kv_indptr[1:] = torch.cumsum(paged_kernel_lens, dim=0)
-    req_pool_indices_cpu = req_pool_indices.cpu().numpy()
-    paged_kernel_lens_cpu = paged_kernel_lens.cpu().numpy()
-    kv_indices = torch.cat(
-        [
-            model_runner.req_to_token_pool.req_to_token[
-                req_pool_indices_cpu[i], : paged_kernel_lens_cpu[i]
-            ]
-            for i in range(batch_size)
-        ],
-        dim=0,
-    ).contiguous()
+    # FIXME (yifan): dirty hack for now. With sequence parallel, we will use flashinfer
+    # paged attn kernel for extend phase as well, and we manually initialize kv_indices
+    # here and avoid write to KV cache. Should revise here after having KV cache support.
+    if forward_mode == ForwardMode.DECODE:
+        req_pool_indices_cpu = req_pool_indices.cpu().numpy()
+        paged_kernel_lens_cpu = paged_kernel_lens.cpu().numpy()
+        kv_indices = torch.cat(
+            [
+                model_runner.req_to_token_pool.req_to_token[
+                    req_pool_indices_cpu[i], : paged_kernel_lens_cpu[i]
+                ]
+                for i in range(batch_size)
+            ],
+            dim=0,
+        ).contiguous()
+    else:
+        kv_indptr = torch.zeros((batch_size + 1,), dtype=torch.int32, device="cuda")
+        kv_indptr[1:] = torch.cumsum(seq_lens - prefix_lens, dim=0)
+        kv_indices = torch.arange(
+            0, torch.sum(seq_lens), dtype=torch.int32, device="cuda"
+        )
```

Comment:
this seems pretty hacky and I don't understand why arange is correct...

Reply:
We will get rid of this magic after switching to use ragged attention for non-causal attention parts.

Reply:
Fixed.